### DISCUSSION TITLE
Solidity教程计划讨论

### DISCUSSION THREAD

**Discussion by AmazingAng** [AmazingAng/WTF-Solidity#3] [General]
## 基础内容（Basic）
constant/immutable （`更新到第14讲中，2022.04.28`）
调用父合约函数 (`添加到第10讲继承中,2022.04.29`)
函数重载overloading（`更新到第15讲中,2022.04.29`）
接收ETH(Fallback, receive)（`更新到第16讲中,2022.05.01`）
发送ETH（`更新到第17讲中,2022.05.04`）
函数输出（`更新到第18讲中,2022.05.04`）
调用其他合约（`更新到第19讲中,2022.05.06`）
状态变量/局部变量/全局变量 (`添加到第4讲中,2022.05.08`)
Enum枚举 (`添加到第2讲中,2022.05.09`)
变量默认值（`更新到第20讲中,2022.05.10`）
Tenary Expression 三元运算符  (`添加到第7讲中,2022.06.04`)

## 进阶内容（Medium）
Call（`更新到第21讲中,2022.05.12`）
DelegateCall（`更新到第22讲中,2022.05.13`）
create contract（`更新到第23讲中,2022.05.14`）
create2（`更新到第24讲中,2022.05.15`）
删除合约（`更新到第25讲中,2022.05.16`）
abi编码解码（`更新到第26讲中,2022.05.20`）
Hash(keccak256)（`更新到第27讲中,2022.05.22`）
函数选择器（selector）（`更新到第28讲中,2022.05.23`）
Try-Catch（`更新到第29讲中,2022.05.25`）
Import（`更新到第30讲中,2022.05.27`）
Staticcal （`比较小众，等有用武之地再加入教程`）

## 应用 (Application)
ERC20 （`更新到第31讲中,2022.06.28`）
水龙头（`更新到第32讲中,2022.06.29`）
空投（`更新到第33讲中,2022.06.30`）
ERC721（`更新到第34讲中,2022.07.03`）
荷兰拍卖（`更新到第35讲中,2022.07.05`）
签名（`更新到第37讲中,2022.07.12`）
Merkle Tree（`更新到第36讲中,2022.07.07`）
NFT交易所（`更新到第38讲中,2022.07.14`）
随机数（`更新到第39讲中,2022.07.19`）
ERC1155（`更新到第40讲中,2022.07.22`）
WETH（`更新到第41讲中,2022.08.03`）
Payment Split（`更新到第42讲中,2022.08.04`）
线性释放（`更新到第43讲中,2022.08.08`）
代币锁（`更新到第44讲中,2022.08.11`）
时间锁（`更新到第45讲中,2022.08.13`）

## Degen
代理合约（`更新到第46讲中,2022.08.18`）
可升级合约（`更新到第47讲中,2022.09.04`）
透明可升级合约（`更新到第48讲中,2022.09.24`）
UUPS合约（`更新到第49讲中,2022.09.26`）
多签（`更新到第50讲中,2022.09.28`）
ERC4626 代币金库 (`更新到第51讲中`)
EIP712签名(`更新到第52讲中`)
ERC20Permit(`更新到第53讲中`)
跨链桥(`更新到第54讲中`)
Multicall(`更新到第55讲中`)
去中心化交易所(`更新到第56讲中`)
Swap
闪电贷(`更新到第57讲中`)
价格预言机
收益农场
链上借贷
去中心化稳定币
以太坊域名
链上投票
?内联汇编

?元交易
?委托投票

## 攻击（Hacks）
重入攻击（`更新到S01中,2022.10.10`）
选择器碰撞（`更新到S02中,2022.10.12`）
中心化风险（`更新到S03中,2022.10.17`）
权限管理漏洞（`更新到S04中,2022.10.19`）
整型溢出（`更新到S05中,2022.10.21`）
签名重放（`更新到S06中,2022.11.02`）
伪随机数漏洞（`更新到S07中,2022.11.10`）
绕过合约长度检验（`更新到S08中,2022.11.12`）
DoS攻击（`更新到S09中,2022.11.13`）
貔貅（`更新到S10中,2022.11.18`）
front-running（`更新到S11中,2022.12.05`）
tx.origin钓鱼攻击（`更新到S12中,2022.12.10`）
Silent Failing Sends and Unchecked-Send attacks （`更新到S13`）
操纵block timestamp（Time Manipulation ）（`更新到S14`）
预言机攻击 （`更新到S15`）
NFT重入攻击（`更新到S16`）
?短地址攻击（Short Address Attack）
?可升级合约漏洞
?DelegateCall漏洞
?自毁
?Gas Griefing attack
?隐藏恶意代码

**Comment by tangminjie** [comment]
内联汇编

**Reply by AmazingAng** [reply]
已加

**Comment by SolidityGo** [comment]
以开发者视角的(非 hack视角), 常写出的 bug 展示以及分析, 对于新手而言很有用, 可以避免踩不必要的坑, 集思广益, 让大家把自己写过的 solidity 比较容易写出来的 bug, 或者和其他语言直观上不太一样的地方, 汇总起来

**Reply by AmazingAng** [reply]
好建议，可以搞个 常见bug 章节？

**Reply by SolidityGo** [reply]
我看行

**Comment by SolidityGo** [comment]
攻击 章节, 可以放上 主流项目的被攻击案例分析, 尤其是最新热门的 攻击分析, 会让学习者更加影响深刻

**Comment by SolidityGo** [comment]
进阶内容（Medium）
增加 `staticcall`,  在合约调用时如何不改变状态的情况下, 调用 `会改变状态的 function`,

**Reply by AmazingAng** [reply]
好建议

**Comment by SolidityGo** [comment]
`攻击（Hacks）`
增加,  
1. `approve` 给可升级合约, 攻击,  example:  `Meerkat`
2. `function` 权限设置错误导致的攻击, 例如 `onlyAdmin` 忘加, 地址普通人可以调用本应 admin 调用的 function
3. 钓鱼链接授权攻击, example: `Opensea` 钓鱼链接, 用户误授权 NFT 被转让 
4. 允许外部函数/合约作为参数传入调用, 导致的攻击,  example:  `O3 swap`

**Reply by AmazingAng** [reply]
好建议

**Comment by EasyChris** [comment]
会增加mev相关的么？ [最大可提取价值 (MEV)](https://ethereum.org/zh/developers/docs/mev/)

例如如何根据aave发起一笔闪电贷等

**Reply by AmazingAng** [reply]
感觉之后可以开一个defi的专题

**Comment by Loius0000** [comment]
请问数值类型quiz最后一题是不是没有正确答案呢？

**Comment by ISheepp** [comment]
OpenZeppelin 源码分析
